sui_core/src/consensus_adapter.rs

```diff
@@ -0,0 +1,178 @@
+                    let digest = Self::hash(&serialized);
+                    self.pending.entry(digest).or_insert_with(VecDeque::new).push_back(replier);
+                },
+
+                Some(output) = self.rx_consensus_output.recv() => {
+                    // Notify the caller that the transaction has been sequenced.
+                    let (outcome, transaction) = output;
+                    let digest = Self::hash(&transaction);
+                    if let Some(repliers) = self.pending.get_mut(&digest) {
+                        if let Some(replier) = repliers.pop_front() {
+                            if replier.send(outcome).is_err() {
+                                debug!("No replier to listen to consensus output {digest}");
+                            }
+                        }
+                    }
+                }
+            }
+        }
+    }
+
+    /// Hash serialized consensus transactions. We do not need specific cryptographic properties except
+    /// only collision resistance.
+    fn hash(serialized: &SerializedConsensusTransaction) -> ConsensusTransactionDigest {
+        let mut hasher = DefaultHasher::new();
+        serialized.hash(&mut hasher);
+        hasher.finish()
+    }
+}
+
+/// Submit Sui certificates to the consensus.
+pub struct ConsensusSubmitter {
+    /// The network address of the consensus node.
+    consensus_address: SocketAddr,
+    /// The network buffer size.
+    buffer_size: usize,
+    /// The Sui committee information.
+    committee: Committee,
+    /// A channel to notify the consensus listener of new transactions.
+    tx_consensus_listener: Sender<ConsensusInput>,
+}
+
+impl ConsensusSubmitter {
+    /// Make a new Consensus submitter instance.
+    pub fn new(
+        consensus_address: SocketAddr,
+        buffer_size: usize,
+        committee: Committee,
+        tx_consensus_listener: Sender<ConsensusInput>,
+    ) -> Self {
+        Self {
+            consensus_address,
+            buffer_size,
+            committee,
+            tx_consensus_listener,
+        }
+    }
+
+    /// Attempt to reconnect with a the consensus node.
+    async fn reconnect(address: SocketAddr, buffer_size: usize) -> SuiResult<TcpDataStream> {
+        transport::connect(address.to_string(), buffer_size)
+            .await
+            .map_err(|e| SuiError::ConsensusConnectionBroken(e.to_string()))
+    }
+
+    /// Submit a transaction to consensus, wait for its processing, and notify the caller.
+    pub async fn submit(&self, certificate: &ConsensusTransaction) -> SuiResult<()> {
+        // Check the Sui certificate (submitted by the user).
+        certificate.check(&self.committee)?;
+
+        // Send certificate to consensus
+        let serialized = serialize_consensus_transaction(certificate);
+        let bytes = Bytes::from(serialized.clone());
+        // TODO [issue #1452]: We are re-creating a connection every time. This is wasteful but does not
```

Comment:
There's a couple other ways to approach this:
1. Use a real RPC like framework that just handles these things for us (e.g. gRPC like we're doing for our other narwhal integrations)
2. store an `Option<TcpStream>` and if its none do the reconnection otherwise use the existing connection.

Reply:
I initially went for option 2 (store an `Option<TcpStream>`) but this function cannot take a mutable reference to self (for reasons linked to authority state). but what francois mentions is a problem I will try to fix it in a separate PR